src/browser-sandbox/pruneNonCriticalCss.js

```diff
@@ -132,103 +135,78 @@ export default function pruneNonCriticalCss ({
     return aboveFold
   }
 
-  function translateSelectorObjectToString (selectorObject) {
-    var parsedSelector = window.csstree.fromPlainObject(selectorObject)
-    var selectorString = window.csstree.translate(parsedSelector)
-    return selectorString
-  }
+  function processCssRules (ast) {
+    console.log('debug: processCssRules BEFORE')
 
-  function translateSelectorListRuleToStrings (selectorListRuleObject) {
-    return selectorListRuleObject.prelude.children.map(
-      translateSelectorObjectToString
-    )
-  }
+    ast = csstree.fromPlainObject(ast)
+
+    csstree.walk(ast, {
+      visit: 'Rule',
+      enter: function (rule, item, list) {
+        // ignore rules inside @keyframes
+        if (
+          this.atrule &&
+          csstree.keyword(this.atrule.name).basename === 'keyframes'
+        ) {
+          return
+        }
 
-  function isCssSelectorRuleCritical (rule) {
-    // check what, if any selectors are found above fold
-    let selectors = []
-    if (rule.prelude.type === 'SelectorList') {
-      selectors = translateSelectorListRuleToStrings(rule)
-    }
-    if (selectors.length === 0) {
-      return false
-    }
+        if (rule.prelude.type !== 'SelectorList') {
+          // remove a rule with bad selector
+          list.remove(item)
+          return
+        }
 
-    // filter out the ones that are not critical
-    const criticalSelectorsMeta = selectors.map((selector, index) => {
-      return {
-        toKeep: isSelectorCritical(selector)
+        // check what, if any selectors are found above fold
+        // filter out the ones that are not critical
+        rule.prelude.children = rule.prelude.children.filter(isSelectorCritical)
+
+        // NOTE: isCssSelectorRuleCritical mutates the rule
+        if (rule.prelude.children.isEmpty()) {
+          list.remove(item)
+        }
       }
     })
-    const criticalSelectorsLength = criticalSelectorsMeta.filter(s => s.toKeep)
-      .length
-    const keepRule = criticalSelectorsLength > 0
-    if (keepRule && criticalSelectorsLength < selectors.length) {
-      // NOTE: mutating the rule here
-      const selectorListFiltered = rule.prelude.children.filter(
-        (rule, index) => {
-          return criticalSelectorsMeta[index].toKeep
-        }
-      )
-      rule.prelude.children = selectorListFiltered
-    }
-
-    return keepRule
-  }
 
-  function isCssRuleCritical (rule) {
-    if (rule.type === 'Rule') {
-      // NOTE: isCssSelectorRuleCritical mutates the rule
-      return isCssSelectorRuleCritical(rule)
-    }
-    /* ==@-rule handling== */
-    /* - Case 0 : Non nested @-rule [REMAIN]
-     (@charset, @import, @namespace)
-     */
-    if (
-      rule.type === 'Atrule' &&
-      (rule.name === 'charset' ||
-        rule.name === 'import' ||
-        rule.name === 'namespace')
-    ) {
-      return true
-    }
+    csstree.walk(ast, {
+      visit: 'Atrule',
```

Comment:
Verified that it slows Penthouse down doing it in this order, so please inverse it.

Reply:
I rely on current test suite, so I tried repeat the logic. As I see now, proposed solution is a little bit different from original one. Not sure we can place it first, because `@media`, `@document` and `@supports` are testing for emptiness. I think at-rule filtering based on name should be in pre-processing and test for empty blocks should be at the ending of post-processing, both outside of sandbox.

Reply:
until at-rule filtering lives in preprocessing (which based on my other comments might end up inside this function anyway), we should just reverse these two to prevent unnecessary processing. The order used previously did not matter as the css was traversed in chronological order.